Albany: Add Albany package.

gahansen opened the pull request:
* Add package.py to support the Albany GitHub project builds.

* Update the Trilinos package.py script to provide additional
  options needed for Albany.

davydden commented on var/spack/repos/builtin/packages/albany/package.py:
```diff
@@ -0,0 +1,128 @@
+    maintainers = ['gahanse']
+
+    # Add proper versions and checksums here.
+    version('master', git='https://github.com/gahansen/Albany.git', branch='master')
```
please call the version `develop`, http://spack.readthedocs.io/en/latest/packaging_guide.html#version-comparison

gahansen replied:
I'll fix...

davydden commented on var/spack/repos/builtin/packages/albany/package.py:
```diff
@@ -0,0 +1,128 @@
+    version('master', git='https://github.com/gahansen/Albany.git', branch='master')
+
+    variant('install',          default=True,
+            description='Enable Install')
```
what does this mean?

gahansen replied:
There is currently an explicit install option (on or off) in Albany's build process. I can eliminate this if I clean that option up on the Albany side. I'll do that and eliminate this option.

gahansen replied:
On further inspection, some Albany users need this option. I'll explicitly set it in the update instead of having it as a variant.

davydden commented on var/spack/repos/builtin/packages/albany/package.py:
```diff
@@ -0,0 +1,128 @@
+            description='Enable LCM_SPECULATIVE')
+    variant('lame',          default=False,
+            description='Enable LAME')
+    variant('edebug',          default=False,
```
elsewhere we call it `debug`.

gahansen replied:
I'll fix...

davydden commented on var/spack/repos/builtin/packages/albany/package.py:
```diff
@@ -0,0 +1,128 @@
+            description='Enable ALBANY_CI')
+    variant('ascr',          default=False,
+            description='Enable ALBANY_ASCR')
+    variant('perf',          default=False,
```
does this run some tests? If so, it's not a variant:

```
spack install -h
--run-tests           run package tests during installation (same as --test=all)
```

gahansen replied:
No - sorry for the confusion. Albany optionally links to a Configuration Interface (CI) that is a GUI that helps users setup the Albany input and execution spec. I could add a better description than "Enable ALBANY_CI" at the expense of making that string longer. Are there limitations on the length of that string (either technical or project policies?)

adamjstewart replied:
There are no limitations on the length of the description. However, if the line goes over 79 characters, flake8 will complain. As long as you keep the entire `variant(...)` on a single line, flake8 will not check that line.

davydden commented on var/spack/repos/builtin/packages/albany/package.py:
```diff
@@ -0,0 +1,128 @@
+    # Add dependencies
+    depends_on('mpi')
+    depends_on('trilinos@develop:~superlu-dist+isorropia+tempus+rythmos+teko+intrepid+intrepid2+minitensor+phal+pnetcdf+nox+piro+rol+shards+stk+superlu')
+    depends_on('trilinos@develop:~superlu-dist+isorropia+tempus+rythmos+teko+intrepid+intrepid2+minitensor+phal+pnetcdf+nox+piro+rol+shards+stk+superlu', when='+lcm')
```
why do you need this one if it's the same as the previous line?

gahansen replied:
I'll fix...

adamjstewart commented on var/spack/repos/builtin/packages/albany/package.py:
```diff
@@ -0,0 +1,128 @@
+       including fluid mechanics, solid mechanics (elasticity and plasticity),
+       ice-sheet flow, quantum device modeling, and many other applications."""
+
+    # Add a proper url for your package's homepage here.
```
You can remove this comment.

gahansen replied:
I'll fix...

adamjstewart commented on var/spack/repos/builtin/packages/albany/package.py:
```diff
@@ -0,0 +1,128 @@
+    homepage = "http://gahansen.github.io/Albany"
+    url      = "https://github.com/gahansen/Albany/tarball/master"
+
+    maintainers = ['gahanse']
```
We usually use GitHub handles in the `maintainers` list so that we can ping people when there are build problems.

gahansen replied:
I'll fix...

adamjstewart commented on var/spack/repos/builtin/packages/albany/package.py:
```diff
@@ -0,0 +1,128 @@
+
+    maintainers = ['gahanse']
+
+    # Add proper versions and checksums here.
```
You can delete this comment.

gahansen replied:
I'll fix...

adamjstewart commented on var/spack/repos/builtin/packages/albany/package.py:
```diff
@@ -0,0 +1,128 @@
+
+        options.extend([
+            '-DALBANY_TRILINOS_DIR:FILEPATH={0}'.format(trilinos_dir),
+            '-DCMAKE_INSTALL_PREFIX:PATH=%s' % self.prefix
```
You don't need this one, Spack already adds it for you in `CMakePackage`.

gahansen replied:
I'll fix...

aprokop commented on var/spack/repos/builtin/packages/trilinos/package.py:
```diff
@@ -147,6 +147,10 @@ class Trilinos(CMakePackage):
             description='Enable Intrepid')
     variant('intrepid2',    default=False,
             description='Enable Intrepid2')
+    variant('minitensor',    default=False,
```
Please add packages in alphabetical order, this makes it easier to maintain.

aprokop commented on var/spack/repos/builtin/packages/trilinos/package.py:
```diff
@@ -322,6 +336,18 @@ def cmake_args(self):
                 'ON' if '+intrepid' in spec else 'OFF'),
             '-DTrilinos_ENABLE_Intrepid2=%s' % (
                 'ON' if '+intrepid2' in spec else 'OFF'),
+            '-DTrilinos_ENABLE_Phalanx=%s' % (
```
Please use alphabetical order here.

gahansen replied:
Will do.

aprokop commented on var/spack/repos/builtin/packages/trilinos/package.py:
```diff
@@ -322,6 +336,18 @@ def cmake_args(self):
+            '-DTrilinos_ENABLE_Rythmos=%s' % (
+                'ON' if '+rythmos' in spec else 'OFF'),
+            '-DTrilinos_ENABLE_Teko=%s' % (
+                'ON' if '+teko' in spec else 'OFF'),
```
Please add checks like `conflict` for configurations you know are incompatible so that they fail in the spack resolution phase and not during Trilinos configure/build. For example, `Teko` has a required dependency on many packages including `Tpetra`, `ML`, etc. Thus, it would be nice to have things like
```python
    conflicts('+piro', when='~tpetra')
```
I understand that this would be time consuming, but I don't see a better solution (would be nice if we could somehow do parts of spack's trilinos package by parsing Trilinos `Dependencies.cmake`).

gahansen replied:
I'll add that dependency. TriBits has an effective dependency system in place, we will need to dig through Dependencies.cmake and duplicate the dependencies in the Spack file for the options that I added. Would you be open to helping me with that?

aprokop replied:
@gahansen Would be glad to help.

gahansen replied:
Thanks! Email is inbound...

aprokop commented:
@gahansen I created a separate Trilinos update PR, that incorporates your changes and includes a bunch of checks for package depencies in #8363. Can you please rebase this PR on top of that and test it out?

gahansen commented:
Commit ba7f3fa replaces the above. @aprokop very kindly adopted my suggested changes to the Trilinos package and submitted PR outlined in #8363 which provides the changes needed that the Albany package requires. This commit contains only the albany/package.py file that builds on that for the new Albany package. I believe I addressed the above comments in this commit. Thanks to everyone who helped with this!

adamjstewart commented:
> I think there might be a priority in spack, with the "@develop" is highest, but the numeric version is preferred over other string identifiers.

Correct. By default, Spack installs the newest numeric release for stability. If you want a different release, you have to explicitly request it.

aprokop commented:
I think you can do something like `depends_on('trilinos@master,develop~superlu...)`. Nalu seems to be doing something similar.

gahansen commented:
Thanks @aprokop! Your syntax suggestion for the Trilinos dependency works great. I added 0f34beb to change this Albany dependency to your suggestion.

aprokop commented:
Just one minor point of confusion for me. @adamjstewart Is `@develop:` same as `@develop`?

adamjstewart commented:
Yes, develop sorts higher than any other version.

aprokop commented:
@adamjstewart Thanks! @gahansen So, you could remove the `:` at the end of `depends_on('trilinos...')` line. Or you could leave it as it is. I have no further comments.

gahansen commented:
@adamjstewart - I'm new to the Spack effort. Do I need to take further action to move forward to a merge?

adamjstewart commented:
I think the only thing left is @aprokop's comment on the unnecessary colon.

adamjstewart commented:
Will merge as soon as @davydden approves.
